source/credentials_provider_ecs.c

```diff
@@ -0,0 +1,667 @@
+    if (!cJSON_IsString(secret_access_key) || (secret_access_key->valuestring == NULL)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "Failed to parse SecretAccessKey from ECS response Json document.");
+        goto parse_error;
+    }
+
+    cJSON *session_token = cJSON_GetObjectItemCaseSensitive(document_root, aws_string_c_str(s_session_token_name));
+    if (!cJSON_IsString(session_token) || (session_token->valuestring == NULL)) {
+        AWS_LOGF_ERROR(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "Failed to parse Token from ECS response Json document.");
+        goto parse_error;
+    }
+
+    cJSON *creds_expiration =
+        cJSON_GetObjectItemCaseSensitive(document_root, aws_string_c_str(s_creds_expiration_name));
+    if (!cJSON_IsString(creds_expiration) || (creds_expiration->valuestring == NULL)) {
+        AWS_LOGF_ERROR(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "Failed to parse Expiration from ECS response Json document.");
+        goto parse_error;
+    }
+
+    /*
+     * Build the credentials
+     */
+    struct aws_byte_cursor access_key_id_cursor = aws_byte_cursor_from_c_str(access_key_id->valuestring);
+    struct aws_byte_cursor secret_access_key_cursor = aws_byte_cursor_from_c_str(secret_access_key->valuestring);
+    struct aws_byte_cursor session_token_cursor = aws_byte_cursor_from_c_str(session_token->valuestring);
+    struct aws_byte_cursor creds_expiration_cursor = aws_byte_cursor_from_c_str(creds_expiration->valuestring);
+
+    if (access_key_id_cursor.len == 0 || secret_access_key_cursor.len == 0 || session_token_cursor.len == 0) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "ECS credentials provider received unexpected credentials response,"
+            " either access key, secret key or token is empty.")
+        goto parse_error;
+    }
+
+    credentials = aws_credentials_new_from_cursors(
+        allocator, &access_key_id_cursor, &secret_access_key_cursor, &session_token_cursor);
+
+    if (credentials == NULL) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "ECS credentials provider failed to allocate memory for credentials.");
+        goto done;
+    }
+
+    if (creds_expiration_cursor.len != 0) {
+        struct aws_date_time expiration;
+        if (aws_date_time_init_from_str_cursor(&expiration, &creds_expiration_cursor, AWS_DATE_FORMAT_ISO_8601) ==
+            AWS_OP_ERR) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Expiration in ECS response Json document is not a valid ISO_8601 date string.");
+            aws_credentials_destroy(credentials);
+            credentials = NULL;
+            goto parse_error;
+        }
+        credentials->expiration_timepoint_seconds = (uint64_t)aws_date_time_as_epoch_secs(&expiration);
+    }
+
+parse_error:
+    aws_raise_error(AWS_AUTH_PROVIDER_PARSER_UNEXPECTED_RESPONSE);
+
+done:
```

Comment:
trivial: maybe throw in a `goto done;` once we know it's all successful, so that we skip over the `parse_error:` section of the code when everything's going ok.

It's not technically wrong to aws_raise_error() when you're returning a successful result in the end, but it's a little weird